targets/TARGET_NXP/TARGET_LPC176X/device/flash_api.c

```diff
@@ -112,42 +112,56 @@ int32_t flash_program_page(flash_t *obj, uint32_t address,
         const uint8_t *data, uint32_t size)
 {
     unsigned long n;
+    const uint32_t copySize = 1024; // should be 256|512|1024|4096
+    uint8_t *alignedData, *source;
+
     // always malloc outside critical section
-    uint8_t *alignedData = malloc(size);
+    alignedData = malloc(size);
```

Comment:
it could be as @davidsaada proposed - smaller program page size if it is defined, or what I was initially thinking, was allocating only `copySize` buffer on heap (because that one is used for flashing). It increases the runtime but making it possible to work for any app (1k of heap should be available). ?

Reply:
Not sure it would help in this case. I suspect that the extra allocation of 1024 bytes already pushes this test over the limit. However, it's worth the try.

Reply:
@0xc0170 @davidsaada In the current code, the reason of the malloc usage is to handle unaligned the data buffer here:

https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_NXP/TARGET_LPC176X/device/flash_api.c#L132-L137

However, there is an alignment check logic here before calling the HAL driver:
https://github.com/ARMmbed/mbed-os/blob/master/drivers/FlashIAP.cpp#L101

I think therefore, the malloc in the flash_program_page() is not required and it should just check the size of flash data.

I am going to modify this and create new commit for review next week.

Reply:
@toyowata You can still check the alignment for the (unlikely) case this code is not called from the driver (or if the driver logic is changed for some reason), and only then perform the malloc + memcpy. In the likely aligned case, you can point to the user buffer.